civis/io/_tables.py

```diff
@@ -782,18 +830,76 @@ def _get_sql_select(table, columns=None):
     return sql
 
 
-def _download_file(url, local_path):
-    response = requests.get(url, stream=True)
-    response.raise_for_status()
+def _get_headers(client, sql, database, credential_id):
+    sql = 'select * from ({}) limit 1'.format(sql)
+    fut = query_civis(sql, database, client=client,
+                      credential_id=credential_id)
+    return fut.result()['result_columns']
 
-    chunk_size = 32 * 1024
-    with open(local_path, 'wb') as fout:
-        chunk = response.iter_content(chunk_size)
-        for lines in chunk:
-            fout.write(lines)
 
+def _decompress_stream(url, buf, chunk_size=32*1024, bytes=True):
+    r = requests.get(url, stream=True)
+    r.raise_for_status()
+
+    d = zlib.decompressobj(zlib.MAX_WBITS | 32)
+    response_iter = r.iter_content(chunk_size)
+    chunk = next(response_iter, '')
+
+    # decompress as stream
+    while chunk or d.unused_data:
+        if d.unused_data:
+            to_decompress = d.unused_data + chunk
+            d = zlib.decompressobj(zlib.MAX_WBITS | 32)
+        else:
+            to_decompress = d.unconsumed_tail + chunk
+        if bytes:
+            buf.write(d.decompress(to_decompress))
+        else:
+            buf.write(d.decompress(to_decompress).decode('utf-8'))
+        chunk = next(response_iter, '')
+
+
+def _download_file(url, local_path, headers, compression):
+    chunk_size = 32 * 1024
 
-def _download_callback(job_id, run_id, client, filename):
+    # if headers are not requested then just write the stream
+    # to the local path since the stream is gzipped
+    if compression == 'gzip' and not headers:
+        response = requests.get(url, stream=True)
+        response.raise_for_status()
+        with open(local_path, 'wb') as fout:
+            shutil.copyfileobj(response.raw, fout, chunk_size)
+
+        return
+
+    # for no compression, decompress the stream
+    if compression == 'none':
+        with open(local_path, 'wb') as fout:
+            fout.write(headers)
+            _decompress_stream(url, fout)
+
+        return
+
+    # if headers are requested or zip compression is requested
+    # first decompress the stream and write headers
+    # then re-compress the file contents
+    with tempfile.NamedTemporaryFile() as temp:
```

Comment:
We had trouble with `NamedTemporaryFile` for Windows users before.  Is there a way to do this without using that function?

https://docs.python.org/3/library/tempfile.html#tempfile.NamedTemporaryFile

Reply:
@keithing yikes! darn windows...

at the bottom of the page it outlines a solution, which is to pass delete=False and then us os.unlink after you're done with the file. still feels more straightforward than tempfile.mkstemp().

also, solution is repeated on SO [https://stackoverflow.com/questions/94153/how-do-i-persist-to-disk-a-temporary-file-using-python](here)

thoughts?

Reply:
You could use a temporary directory instead. We do that in several other places in the client. You need to import the version from `compat`; the `TemporaryDirectory` isn't in Python 2.

Reply:
great. that looks better.

also, i think i need to fix the use of `TemporaryDirectory` in the multipart-upload PR then
https://github.com/civisanalytics/civis-python/blob/master/civis/io/_files.py#L163

Reply:
Is there a way to avoid the second pass over the data in the first place?

Reply:
not really... parallel unload will only return no compression, gzip or bzip2.

if we want to be able to return gzip or zip files then we'll have to download the file while decompressing the stream, write headers, then zip/gzip it.

the issue is that we're including headers. i'm not sure how to get around that.

Reply:
You can concatenate gzipped files into a larger gzipped file.

Reply:
- removed tempfile and switched to civis.compat.TemporaryDirectory
- simplified code for gzips by just writing headers as gzipped bytes since gzipped files can be concatenated